File: src/main/java/core/procedures/hashmaps/Merge.java
package core.procedures.hashmaps;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMVoid;

import java.util.HashMap;
import java.util.Map;

public final class Merge extends AFn {

  public Merge() {
    super(new FnArgsBuilder().min(0).rest(Map.class).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "merge";
  }

  @Override
  public Object apply(Object... args) {
    if (args.length == 0) {
      return SCMVoid.VOID;
    }
    Map<Object, Object> result = new HashMap<>();
    for (Object m : args) {
      result.putAll((Map)m);
    }
    return result;
  }
}

File: src/main/java/core/procedures/hashmaps/Merge.kt
package core.procedures.hashmaps

import core.procedures.AFn
import core.scm.MutableHashmap

class Merge : AFn<Any?, Map<*, *>?>(name = "merge", isPure = true, restArgsType = Map::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when {
        args.isEmpty() -> null
        else -> MutableHashmap<Any?, Any?>().apply { args.forEach { putAll(it as Map<*, *>) } }
    }
}
